changelog_unreleased/cli/12800.md

```diff
@@ -0,0 +1,24 @@
+#### [HIGILIGHT]Add `--cache` and `--cache-strategy` CLI option (#12800 by @sosukesuzuki)
+
+Two new CLI options have been added for a caching system similar to [ESLint's one](https://eslint.org/docs/user-guide/command-line-interface#caching).
+
+##### `--cache`
+
+If this option is enabled, the following values are used as cache keys and the file is formatted only if one of them is changed.
+
+- Options
+- Node.js version
+- (if `--cache-strategy` is `metadata`) file metadata, such as timestamps
+- (if `--cache-strategy` is `content`) content of the file
+
+```bash
+prettier --write --cache src
+```
+
+##### `--cache-strategy`
+
+Strategy for the cache to use for detecting changed files. Can be either `metadata` or `content`. If no strategy is specified, `metadata` will be used.
+
+```bash
+prettier --write --cache --cache-strategy content src
+```
```

Comment:
Should we help users choosing by having some extra text like ESLint?

https://eslint.org/docs/user-guide/command-line-interface#--cache-strategy

> The content strategy can be useful in cases where the modification time of your files change even if their contents have not. For example, this can happen during git operations like git clone because git does not track file modification time.

Is another use case CI?

Btw, what is the reason for having both metadata and content? Why not always content? Is it performance? What’s the performance difference? It feels like there’s a tradeoff here that we want to document.

Edit: Oops, I meant to comment on the CLI docs, not the changelog.

Reply:
We originally did not intend to add the `--cache-strategy` option and always intended to use `metadata`. However, following a comment from @7rulnik ( https://github.com/prettier/prettier/pull/12800#discussion_r878044023 ), we added the `--cache-strategy` option for CI use cases so that `content` can be used.

I had thought there was a tradeoff between `metadata` and `content` as follows:

|        | `metadata`  | `content`   |
| ------ | ----------- | ----------- |
| for CI | :no_good:   | :ok_person: |
| perf   | :ok_person: | :no_good:   |



However, following your comment, I investigated and found that there is actually not much difference in performance between the two.

***

`--cache-strategy=metadata`:

```
$ ./bin/prettier.js . "!test*" --check --write --cache
Checking formatting...
All matched files use Prettier code style!
✨  Done in 16.51s.

$ ./bin/prettier.js . "!test*" --check --write --cache
Checking formatting...
All matched files use Prettier code style!
✨  Done in 6.56s.
```

`--cache-strategy=content`:

```
$ ./bin/prettier.js . "!test*" --check --write --cache --cache-strategy=content
Checking formatting...
All matched files use Prettier code style!
✨  Done in 16.46s.

$ ./bin/prettier.js . "!test*" --check --write --cache --cache-strategy=content
Checking formatting...
All matched files use Prettier code style!
✨  Done in 6.81s.
```

***

From this result I think we can remove the `--cache-strategy` option and always use `content`.

What do you think? @7rulnik, @fisker

Reply:
Timestamp should be more faster than content strategy, it is very strange that we have almost the same indicators, can you use https://benchmarkjs.com/ and calculate mean? Maybe we have a bug somewhere?

Reply:
I created following script:

<details>
  <summary>test.benchmark.js</summary>
  
```js
const Benchmark = require("benchmark");
const { execaSync } = require("./vendors/execa");

const runWithMetadata = () =>
  execaSync("./bin/prettier.js", [".", "!test*", "--check", "--cache"]);

const runWithContent = () =>
  execaSync("./bin/prettier.js", [
    ".",
    "!test*",
    "--check",
    "--cache",
    "--cache-strategy",
    "content",
  ]);

const clearCache = () =>
  execaSync("rm", ["-f", "./node_modules/.cache/prettier/.prettier-cache"]);

function runSuite01() {
  const suite01 = new Benchmark.Suite();

  clearCache();
  runWithMetadata();

  suite01
    .add("--cache-strategy=metadata", function () {
      runWithMetadata();
    })
    .on("cycle", function (event) {
      console.log(String(event.target));
    })
    .run();
}

function runSuite02() {
  const suit02 = new Benchmark.Suite();

  clearCache();
  runWithContent();

  suit02
    .add("--cache-strategy=content", function () {
      runWithContent();
    })
    .on("cycle", function (event) {
      console.log(String(event.target));
    })
    .run();
}

runSuite01();
runSuite02();

```

</details>

The results were as follows:

```
--cache-strategy=metadata x 0.30 ops/sec ±6.05% (5 runs sampled)
--cache-strategy=content x 0.30 ops/sec ±0.94% (5 runs sampled)
```

I ran several other benchmarks and found no significant differences. This is strange.

But, ESLint cache (Almost the same mechanism as our cache) with similar results.

***

**ESLint with `--cache-strategy=metadata`:**

```
$ cross-env EFF_NO_LINK_RULES=true eslint . --format friendly --cache --cache-strategy=metadata
✨  Done in 16.27s.

$ cross-env EFF_NO_LINK_RULES=true eslint . --format friendly --cache --cache-strategy=metadata
✨  Done in 3.05s.
```

**ESLint with `--cache-strategy=content`:**

```
$ cross-env EFF_NO_LINK_RULES=true eslint . --format friendly --cache --cache-strategy=content
✨  Done in 13.69s.

$ cross-env EFF_NO_LINK_RULES=true eslint . --format friendly --cache --cache-strategy=content
✨  Done in 2.96s.
```

***

I ran the benchmark using Prettier's source code this time, but the results may be different for projects where each file is more huge.

Reply:
(Even in the Babel source code, there was little difference in speed between `metadata` and `content`.)

Reply:
Based on benchmark cold start is faster for `metadata`, but yes, it can be different for projects, not sure how often big projects run prettier on whole project (where `metadata` make sense for perf reasons), but maybe we should keep both values (i.e. `content` and `metadata`)

Reply:
To maintain both `--cache-strategy=content` and `--cache-strategy=metadata` isn't hard. So I think we can keep both.

Reply:
But should we default to content?

How big does a project need to be for metadata to make sense?

Reply:
I created new benchmark script https://gist.github.com/sosukesuzuki/ded586261381a6261e02a69b188a8e44

Results ware follows:

```
======= 100 lines, --cache-strategy=metadata, 0ms ============
======= 1000 lines, --cache-strategy=metadata, 1ms ============
======= 10000 lines, --cache-strategy=metadata, 1ms ============
======= 100000 lines, --cache-strategy=metadata, 0ms ============
======= 100 lines, --cache-strategy=content, 17ms ============
======= 1000 lines, --cache-strategy=content, 12ms ============
======= 10000 lines, --cache-strategy=content, 13ms ============
======= 100000 lines, --cache-strategy=content, 12ms ============
```

This benchmark script measures execution time when cache is enabled. There is only one file of interest, and each test combines the number of lines in the file with the cache strategy.

According to this, performance still seems to be better with `--cache-strategy=metadata`. I think the reason why the previously posted benchmarks did not show a valid difference is probably due to overhead in areas other than formatting.

Reply:
Saving 10-20 ms does not convince me that the option is worth it, but if you think we should have it that’s ok with me.

But I definitely think we should default to content.

Reply:
I've updated the documentation to use `content` by default. For normal projects, there is no performance difference, and `content` is more convenient.